### PR TITLE
Task 5

### PR DESCRIPTION
**Pull request by programister42**
**Import Service API:** https://qdde42rit0.execute-api.us-east-1.amazonaws.com/prod
**FE PR:** https://github.com/programister42/nodejs-aws-shop-react/pull/4
**FE Deploy**: https://d114k010gapdaa.cloudfront.net/admin/products

**Self-evaluation**: all tasks completed - 100/100

### CODE REVIEW

File: import-service/lambdas/import-products-file.ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
+import { PutObjectCommand, S3Client } from "@aws-sdk/client-s3";
+import { getSignedUrl } from "@aws-sdk/s3-request-presigner";
+import type { APIGatewayProxyHandler } from "aws-lambda";
+
+const BUCKET_NAME = process.env.BUCKET_NAME;
+const BUCKET_REGION = process.env.BUCKET_REGION;
+const s3Client = new S3Client({ region: BUCKET_REGION });
```
Comment (hazardsoft) [inline import-service/lambdas/import-products-file.ts:7]: Is `BUCKET_REGION` required?

File: import-service/lambdas/import-file-parser.ts [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+import type { S3Handler } from "aws-lambda";
+import * as csv from "csv-parser";
+
+const BUCKET_REGION = process.env.BUCKET_REGION;
+const s3Client = new S3Client({ region: BUCKET_REGION });
+
+export const importFileParser: S3Handler = async (event) => {
+	console.log("importFileParser event", event);
+
+	for (const record of event.Records) {
```
Comment (hazardsoft) [inline import-service/lambdas/import-file-parser.ts:17]: This is the 1st implementation I see that cares about multiple records in an event! Nice touch!

File: import-service/lambdas/import-file-parser.ts [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+		try {
+			const { Body } = await s3Client.send(
+				new GetObjectCommand({
+					Bucket: bucket.name,
+					Key: object.key,
+				}),
+			);
+			const readableStream = Body as Readable;
+
+			readableStream
```
Comment (hazardsoft) [inline import-service/lambdas/import-file-parser.ts:29]: Stream reading is not wrapped in a promise so it seems that handler will return earlier than the stream ends.

File: test/import-service/lambdas/import-file-parser.test.ts [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+	CopyObjectCommand,
+	DeleteObjectCommand,
+	GetObjectCommand,
+	S3Client,
+} from "@aws-sdk/client-s3";
+import type { StreamingBlobPayloadOutputTypes } from "@smithy/types";
+import type { Context, S3Event } from "aws-lambda";
+import { mockClient } from "aws-sdk-client-mock";
+import { importFileParser } from "../../../import-service/lambdas/import-file-parser";
+import * as csv from "../../../import-service/lambdas/node_modules/csv-parser";
```
Comment (hazardsoft) [inline test/import-service/lambdas/import-file-parser.test.ts:12]: invalid import

File: test/import-service/lambdas/import-file-parser.test.ts [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+jest.mock("../../../import-service/lambdas/node_modules/csv-parser");
+
+const s3Mock = mockClient(S3Client);
+
+describe("importFileParser", () => {
```
Comment (hazardsoft) [inline test/import-service/lambdas/import-file-parser.test.ts:18]: Tests fail with the following reasons:
1. Cannot log after tests are done. Did you forget to wait for something async in your test?

File: import-service/import-service-stack.ts [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+				cors: [
+					{
+						allowedOrigins: ["*"],
+						allowedMethods: [aws_s3.HttpMethods.PUT],
+						allowedHeaders: ["*"],
+					},
+				],
+			},
+		);
+		bucket.addLifecycleRule({
```
Comment (hazardsoft) [inline import-service/import-service-stack.ts:34]: Was not part of the requirements, please elaborate the use of lifecycle here.

**Review by hazardsoft — CHANGES_REQUESTED** [review]
`import-file-parser.ts` implementation seems not to wait the end of a stream and returns early (tests are failing as well for the same reason).

File: import-service/lambdas/import-file-parser.ts (hunk 3 above)
Comment (programister42) [reply import-service/lambdas/import-file-parser.ts]: Thank you for noticing and requesting changes!
[fix: fix importFileParser lambda](https://github.com/programister42/nodejs-aws-shop-backend/pull/3/commits/efd402e3f6154276e8a5e890acfbafb7b878e187)

File: test/import-service/lambdas/import-file-parser.test.ts (hunk 4 above)
Comment (programister42) [reply test/import-service/lambdas/import-file-parser.test.ts]: Refactored: [refactor: rewrite importFileParser test](https://github.com/programister42/nodejs-aws-shop-backend/pull/3/commits/a84a35aac654826692773b1d26b8ba720fcda473)

File: test/import-service/lambdas/import-file-parser.test.ts (hunk 5 above)
Comment (programister42) [reply test/import-service/lambdas/import-file-parser.test.ts:17]: Refactored: [refactor: rewrite importFileParser test](https://github.com/programister42/nodejs-aws-shop-backend/pull/3/commits/a84a35aac654826692773b1d26b8ba720fcda473)

File: import-service/import-service-stack.ts (hunk 6 above)
Comment (programister42) [reply import-service/import-service-stack.ts:34]: It was mentioned in the module materials, so I added it

**Pull request merged**